qiskit/circuit/library/arithmetic/adders/ripple_carry_adder.py

```diff
@@ -0,0 +1,121 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute the sum of two qubit registers using ripple-carry approach."""
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister, AncillaRegister
+
+from .adder import Adder
+
+
+class RippleCarryAdder(Adder):
+    r"""A ripple-carry circuit to perform in-place addition on two qubit registers.
+
+    As an example, a ripple-carry adder circuit that performs addition on two 3-qubit sized
+    registers is as follows:
+
+    .. parsed-literal::
+
+                ┌──────┐┌──────┐                     ┌──────┐┌──────┐
+           a_0: ┤0     ├┤2     ├─────────────────────┤2     ├┤0     ├
+                │      ││      │┌──────┐     ┌──────┐│      ││      │
+           a_1: ┤      ├┤0     ├┤2     ├─────┤2     ├┤0     ├┤      ├
+                │      ││      ││      │     │      ││      ││      │
+           a_2: ┤      ├┤  MAJ ├┤0     ├──■──┤0     ├┤  UMA ├┤      ├
+                │      ││      ││      │  │  │      ││      ││      │
+           b_0: ┤1 MAJ ├┤      ├┤  MAJ ├──┼──┤  UMA ├┤      ├┤1 UMA ├
+                │      ││      ││      │  │  │      ││      ││      │
+           b_1: ┤      ├┤1     ├┤      ├──┼──┤      ├┤1     ├┤      ├
+                │      │└──────┘│      │  │  │      │└──────┘│      │
+           b_2: ┤      ├────────┤1     ├──┼──┤1     ├────────┤      ├
+                │      │        └──────┘  │  └──────┘        │      │
+         cin_0: ┤2     ├──────────────────┼──────────────────┤2     ├
+                └──────┘                ┌─┴─┐                └──────┘
+        cout_0: ────────────────────────┤ X ├────────────────────────
+                                        └───┘
```

Comment:
If you put `cin_0` at the top then the circuit becomes slightly nicer (from my implementation [here](https://github.com/ajavadia/qiskit-terra/blob/Demo/demo/Circuit%20Library%20Demo.ipynb))

```
        ┌──────┐                                                     ┌──────┐
 cin_0: ┤0     ├─────────────────────────────────────────────────────┤0     ├
        │      │┌──────┐                                     ┌──────┐│      │
   a_0: ┤2     ├┤0     ├─────────────────────────────────────┤0     ├┤2     ├
        │      ││      │┌──────┐                     ┌──────┐│      ││      │
   a_1: ┤      ├┤2     ├┤0     ├─────────────────────┤0     ├┤2     ├┤      ├
        │  maj ││      ││      │┌──────┐     ┌──────┐│      ││      ││  uma │
   a_2: ┤      ├┤      ├┤2     ├┤0     ├─────┤0     ├┤2     ├┤      ├┤      ├
        │      ││  maj ││      ││      │     │      ││      ││  uma ││      │
   a_3: ┤      ├┤      ├┤      ├┤2     ├──■──┤2     ├┤      ├┤      ├┤      ├
        │      ││      ││  maj ││      │  │  │      ││  uma ││      ││      │
   b_0: ┤1     ├┤      ├┤      ├┤      ├──┼──┤      ├┤      ├┤      ├┤1     ├
        └──────┘│      ││      ││  maj │  │  │  uma ││      ││      │└──────┘
   b_1: ────────┤1     ├┤      ├┤      ├──┼──┤      ├┤      ├┤1     ├────────
                └──────┘│      ││      │  │  │      ││      │└──────┘        
   b_2: ────────────────┤1     ├┤      ├──┼──┤      ├┤1     ├────────────────
                        └──────┘│      │  │  │      │└──────┘                
   b_3: ────────────────────────┤1     ├──┼──┤1     ├────────────────────────
                                └──────┘┌─┴─┐└──────┘                        
cout_0: ────────────────────────────────┤ X ├────────────────────────────────
                                        └───┘                                
```

Reply:
Yeah, but we should be consistent with where we place the bits such that we can use these adders consistently. If we don't put all auxiliary qubits (like cin) at the bottom, users must be careful of how to compose the different adders. If we always keep the same structure: `a - b - carry (cout) - auxiliaries (cin)` then it's easier to correctly compose them 🙂

Reply:
sure

Reply:
Yes but the placement of bits has value significance. If you look at the list in circuit.qubits, the 0th qubit is the least significant and the Nth qubit is the most significant. This becomes important in statevector and unitary representations of the circuit. So in an adder, carry_in is the least significant bit, and carry_out is the most significant one. 

So I agree with consistency of bit orders across adders, but doesn't this order make more sense: c_in, a, b, c_out?
Are you saying some adders don't have carry_in? My question would be why not? Because that's already a source of inconsistency among adders, and users need to look at when to allocate ancilla for the carry_in/carry_out, and when not.

Reply:
Maybe I'm a bit confused about what c_in means. I used it as analogue for ancilla bits, is that correct? 
But e.g. the QFT adder don't have any c_in, and the VBE ripple carry adder has `n - 1` (for adding `n` bit numbers).

Reply:
But yeah the adders are not consistent in the number of ancillas they need, which is why we had the two register they all share (one for `a` and one for `b`) at the top. Then you can always use the adders in the same fashion:
```python 
adder = # any adder with n bits
adder.compose(a_circuit, qubits=range(n))
adder.compose(b_circuit, qubits=range(n+1, 2*n))
```
and not have to worry about how many ancillas come first 🙂

Reply:
Oh I agree with general ancilla to come last. That is scratch space for computation and is not part of the I/O of the adder. But `c_in` and `c_out` are "carry in" and "carry out" bits of the adder. They have semantic meaning, i.e. if you think of the adder as a black box that accepts two n-bit registers "a" and "b" and sums them up, in general this could produce a final `c_out` bit if the result does not fit into n bits. Or you could invoke this adder black box by not only passing it "a" and "b", but also some extra bit that came from a previous stage of the computation and that's `c_in`.

Does this make sense? @boschmitt might have some thoughts here as well since he's worked with adders. 

I think QFT adder can be made to have `c_in`.

Reply:
I see, so here's one thing I didn't get: `c_in` is required if you want to chain additions. But assume you want to chain a lot of additions, then you might still run out of bits, right? E.g. if you want to add four 1-bit numbers: 1 + 1 + 1 + 1. Then the first addition gives you 10, the second 11 and the last 00? 

That's why I was thinking we just avoid `c_in` altogether and the user must make sure the registers have correct sizes. But if it is custom to have `c_in` we can also add it first of course (maybe with an option to not have it for e.g. the QFT adder).

Reply:
Summarizing our discussion: We have three types of adders, "full" (cin at top, cout at bottom), "half" (no cin, cout at bottom) and "fixed" (no cin, no cout).